mpi-derive/src/lib.rs

```diff
@@ -43,7 +43,7 @@ fn equivalence_for_tuple_field(type_tuple: &syn::TypeTuple) -> TokenStream2 {
 fn equivalence_for_array_field(type_array: &syn::TypeArray) -> TokenStream2 {
     let ty = equivalence_for_type(&type_array.elem);
     let len = &type_array.len;
-    quote! { &::mpi::datatype::UncommittedUserDatatype::contiguous(#len, &#ty) }
+    quote! { &::mpi::datatype::UncommittedUserDatatype::contiguous(#len as i32, &#ty) }
```

Comment:
Could you do something like this instead:

```suggestion
    quote! { &::mpi::datatype::UncommittedUserDatatype::contiguous(#len.try_into().expect(&format!("rsmpi derive: Array size too large for MPI_Datatype: {}", #len)), &#ty) }
```

It's unlikely anybody would ever have an array size that big, but might as well fail loudly if they do.

Reply:
I did something similar. I used a static `&str` instead of a format invocation because I didn't want to make the allocation from `format` each time `equivalent_datatype` is called. Otherwise it has the same effect.

Reply:
I would guess that the actual allocation would be elided except in failure cases since it can be statically determined if the constant value is convertible to an `i32`. But that's fine - looks good!